mem_context.go

```diff
@@ -122,6 +122,18 @@ func NewMemUsageContext(
 	}
 }
 
+// MemUsageLimitExceeded ensures a limit function is defined and checks against the limit. If limit is breached
+// it will return true, if the limit check is undefined it will return an error
+func (m *MemUsageContext) MemUsageLimitExceeded(memUsage uint64) (bool, error) {
+	if m.MemUsageExceedsLimit == nil {
```

Comment:
do we need to perform this check in tight loops? I would rather not if possible. Wouldn't we be able to check this in advance?

Reply:
The var is an exported pointer structure so anyone could change it at any point. I think if we determined that the only ever definition of this function would be:
```go
if memUsage > m.MemLimit {
return err
}
```
then we can remove these checks and clean things up a bit. Happy to do so if everyone is on board.

Reply:
Chatted with @Gabri3l offline and he was fine with moving forward this approach as long as the limit field was not exported. It has been refactored.